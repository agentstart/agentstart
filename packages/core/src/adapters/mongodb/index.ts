/* agent-frontmatter:start
AGENT: MongoDB adapter entry
PURPOSE: Re-export the MongoDB database adapter implementation
USAGE: import { mongodbAdapter } from "@agent-stack/core/adapters/mongodb"
EXPORTS: mongodbAdapter
FEATURES:
  - Shields consumers from folder layout
  - Keeps MongoDB adapter API discoverable
SEARCHABLE: mongodb adapter entry, mongo integration
agent-frontmatter:end */

export * from "./mongodb-adapter";
