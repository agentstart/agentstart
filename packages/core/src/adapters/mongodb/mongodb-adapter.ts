/* agent-frontmatter:start
AGENT: MongoDB adapter
PURPOSE: Provide a database adapter implementation backed by MongoDB
USAGE: mongodbAdapter(db, { provider: "mongodb" })
EXPORTS: mongodbAdapter
FEATURES:
  - Translates normalized filters into MongoDB query documents
  - Supports projections, sorting, and pagination
  - Returns updated documents where supported
SEARCHABLE: mongodb adapter, database adapter, document integration
agent-frontmatter:end */

import { AgentStackError } from "@agent-stack/errors";
import type {
  Db,
  Document,
  Filter,
  FindOneAndDeleteOptions,
  FindOneAndUpdateOptions,
  FindOptions,
  InsertOneResult,
  DeleteResult as MongoDeleteResult,
  UpdateResult as MongoUpdateResult,
  UpdateFilter,
} from "mongodb";

import {
  type AdapterCountArgs,
  type AdapterCreateArgs,
  type AdapterDebugLogger,
  type AdapterDeleteArgs,
  type AdapterFindManyArgs,
  type AdapterFindOneArgs,
  type AdapterUpdateArgs,
  type AdapterUpdateManyArgs,
  type AdapterUpsertArgs,
  createAdapterFactory,
} from "../create-database-adapter";
import {
  type AdapterWhereCondition,
  applyInputTransforms,
  camelToSnake,
  createDebugLoggerHook,
  ensureArrayValue,
  mapSelectToObject,
  normalizeSortInput,
  normalizeWhereInput,
  pluralizeModel,
  splitWhereConditions,
} from "../shared";

export interface MongoAdapterUserOptions {
  usePlural?: boolean;
  camelCase?: boolean;
  collectionMap?: Record<string, string>;
  debug?:
    | boolean
    | {
        logger?: AdapterDebugLogger;
      };
}

interface MongoAdapterResolvedOptions extends MongoAdapterUserOptions {
  db: Db;
}

function escapeRegExp(input: string): string {
  return input.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function buildFieldCondition(
  condition: AdapterWhereCondition,
  fieldNormalizer: (field: string) => string,
): Filter<Document> {
  // MongoDB relies on operator-prefixed objects; perform the mapping here so
  // adapters share the same where representation.
  const fieldName = fieldNormalizer(condition.field);
  switch (condition.operator) {
    case "ne":
      return { [fieldName]: { $ne: condition.value } };
    case "lt":
      return { [fieldName]: { $lt: condition.value } };
    case "lte":
      return { [fieldName]: { $lte: condition.value } };
    case "gt":
      return { [fieldName]: { $gt: condition.value } };
    case "gte":
      return { [fieldName]: { $gte: condition.value } };
    case "in": {
      const values = ensureArrayValue(condition);
      return { [fieldName]: { $in: values } };
    }
    case "not_in": {
      const values = ensureArrayValue(condition);
      return { [fieldName]: { $nin: values } };
    }
    case "contains": {
      const pattern = `.*${escapeRegExp(String(condition.value))}.*`;
      return { [fieldName]: { $regex: new RegExp(pattern) } };
    }
    case "starts_with": {
      const pattern = `^${escapeRegExp(String(condition.value))}`;
      return { [fieldName]: { $regex: new RegExp(pattern) } };
    }
    case "ends_with": {
      const pattern = `${escapeRegExp(String(condition.value))}$`;
      return { [fieldName]: { $regex: new RegExp(pattern) } };
    }
    case undefined:
    case "eq":
      return { [fieldName]: condition.value };
    default: {
      const exhaustive: never = condition.operator;
      throw new AgentStackError(
        "DB_ADAPTER_INVALID_WHERE",
        `Unsupported operator "${String(exhaustive)}" supplied for MongoDB adapter.`,
      );
    }
  }
}

function buildMongoFilter(
  conditions: AdapterWhereCondition[],
  fieldNormalizer: (field: string) => string,
): Filter<Document> {
  if (conditions.length === 0) {
    return {};
  }

  const { andConditions, orConditions } = splitWhereConditions(conditions);

  if (andConditions.length === 1 && orConditions.length === 0) {
    return buildFieldCondition(andConditions[0]!, fieldNormalizer);
  }

  const filter: Filter<Document> = {};
  if (andConditions.length > 0) {
    filter.$and = andConditions.map((condition) =>
      buildFieldCondition(condition, fieldNormalizer),
    );
  }
  if (orConditions.length > 0) {
    filter.$or = orConditions.map((condition) =>
      buildFieldCondition(condition, fieldNormalizer),
    );
  }
  return filter;
}

function convertProjection(
  model: string,
  select: string[] | undefined,
  normalizeFieldName: (model: string, field: string) => string,
): FindOptions<Document>["projection"] {
  // Mongo projections expect a document of `{ field: 1 }` pairs.
  return mapSelectToObject(select, model, normalizeFieldName, () => 1);
}

function buildSortDocument(
  model: string,
  sort: AdapterFindManyArgs["sortBy"],
  normalizeFieldName: (model: string, field: string) => string,
): Document | undefined {
  const [primarySort] = normalizeSortInput(sort);
  if (!primarySort?.field) {
    return undefined;
  }
  return {
    [normalizeFieldName(model, primarySort.field)]:
      primarySort.direction === "desc" ? -1 : 1,
  };
}

const baseMongoAdapter = createAdapterFactory<MongoAdapterResolvedOptions>({
  hooks: {
    logger: createDebugLoggerHook<MongoAdapterResolvedOptions>("mongodb"),
    normalizeModelName: ({ options, model }) => {
      const mapped = options.collectionMap?.[model];
      if (mapped) {
        return mapped;
      }
      return options.usePlural ? pluralizeModel(model) : model;
    },
    normalizeFieldName: ({ options, field }) =>
      options.camelCase ? camelToSnake(field) : field,
  },
  config: (options) => ({
    adapterId: "mongodb",
    adapterName: "MongoDB",
    usePlural: options.usePlural ?? false,
    supportsJSON: true,
    supportsDates: true,
    supportsBooleans: true,
    supportsNumericIds: false,
  }),
  adapter: ({
    options,
    wrapOperation,
    normalizeModelName,
    normalizeFieldName,
    getFieldAttributes,
  }) => {
    const { db } = options;

    const buildFilter = (model: string, where: unknown): Filter<Document> =>
      buildMongoFilter(normalizeWhereInput(where), (field) =>
        normalizeFieldName(model, field),
      );

    return {
      create: wrapOperation<AdapterCreateArgs, unknown>(
        "create",
        async ({ model, data, allowId }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const prepared = applyInputTransforms({
            action: "create",
            model,
            data,
            allowId,
            getFieldAttributes,
          }) as Document;
          const result: InsertOneResult<Document> =
            await collection.insertOne(prepared);
          const insertedId = result.insertedId;
          if (!insertedId) {
            return prepared;
          }
          const inserted = await collection.findOne({ _id: insertedId });
          return inserted ?? { ...prepared, _id: insertedId };
        },
      ),
      upsert: wrapOperation<AdapterUpsertArgs, unknown>(
        "upsert",
        async ({ model, where, create, update, select }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const preparedCreate = applyInputTransforms({
            action: "create",
            model,
            data: create,
            getFieldAttributes,
          }) as Document;
          const preparedUpdate = applyInputTransforms({
            action: "update",
            model,
            data: update,
            getFieldAttributes,
          }) as Document;
          const updateDoc: UpdateFilter<Document> = {
            $set: preparedUpdate,
            ...(Object.keys(preparedCreate).length > 0
              ? { $setOnInsert: preparedCreate }
              : {}),
          };
          const projection = convertProjection(
            collectionName,
            select,
            normalizeFieldName,
          );
          const options: FindOneAndUpdateOptions = {
            upsert: true,
            returnDocument: "after",
            projection,
          };
          const result = await collection.findOneAndUpdate(
            filter,
            updateDoc,
            options,
          );
          if (result?.value) {
            return result.value;
          }
          return collection.findOne(filter, { projection });
        },
      ),
      findOne: wrapOperation<AdapterFindOneArgs, unknown | null>(
        "findOne",
        async ({ model, where, select }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const projection = convertProjection(
            collectionName,
            select,
            normalizeFieldName,
          );
          return collection.findOne(filter, { projection });
        },
      ),
      findMany: wrapOperation<AdapterFindManyArgs, unknown[]>(
        "findMany",
        async ({ model, where, sortBy, limit, offset, select }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const projection = convertProjection(
            collectionName,
            select,
            normalizeFieldName,
          );
          const cursor = projection
            ? collection.find(filter, { projection })
            : collection.find(filter);
          const sortDocument = buildSortDocument(
            collectionName,
            sortBy,
            normalizeFieldName,
          );
          if (sortDocument) {
            cursor.sort(sortDocument);
          }
          if (typeof offset === "number") {
            cursor.skip(offset);
          }
          if (typeof limit === "number") {
            cursor.limit(limit);
          }
          return cursor.toArray();
        },
      ),
      count: wrapOperation<AdapterCountArgs, number>(
        "count",
        async ({ model, where }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          return collection.countDocuments(filter);
        },
      ),
      update: wrapOperation<AdapterUpdateArgs, unknown>(
        "update",
        async ({ model, where, update }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const preparedUpdate = applyInputTransforms({
            action: "update",
            model,
            data: update,
            getFieldAttributes,
          }) as Document;
          const updateDoc: UpdateFilter<Document> = {
            $set: preparedUpdate,
          };
          const options: FindOneAndUpdateOptions = {
            returnDocument: "after",
          };
          const result = await collection.findOneAndUpdate(
            filter,
            updateDoc,
            options,
          );
          if (!result) {
            return null;
          }
          return result.value ?? null;
        },
      ),
      updateMany: wrapOperation<AdapterUpdateManyArgs, number>(
        "updateMany",
        async ({ model, where, update }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const preparedUpdate = applyInputTransforms({
            action: "update",
            model,
            data: update,
            getFieldAttributes,
          }) as Document;
          const updateDoc: UpdateFilter<Document> = {
            $set: preparedUpdate,
          };
          const result: MongoUpdateResult = await collection.updateMany(
            filter,
            updateDoc,
          );
          return result.modifiedCount ?? 0;
        },
      ),
      delete: wrapOperation<AdapterDeleteArgs, unknown>(
        "delete",
        async ({ model, where }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const options: FindOneAndDeleteOptions = {};
          const result = await collection.findOneAndDelete(filter, options);
          if (!result) {
            return null;
          }
          return result.value ?? null;
        },
      ),
      deleteMany: wrapOperation<AdapterDeleteArgs, number>(
        "deleteMany",
        async ({ model, where }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const result: MongoDeleteResult = await collection.deleteMany(filter);
          return result.deletedCount ?? 0;
        },
      ),
      options,
    };
  },
});

export function mongodbAdapter(db: Db, options: MongoAdapterUserOptions = {}) {
  return baseMongoAdapter({
    ...options,
    db,
  });
}
