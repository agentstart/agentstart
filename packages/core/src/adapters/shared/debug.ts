/* agent-frontmatter:start
AGENT: Adapter debug helpers
PURPOSE: Provide shared debug logging hook for adapters
USAGE: createDebugLoggerHook("prisma")
EXPORTS: createDebugLoggerHook
FEATURES:
  - Handles boolean or custom logger debug options
  - Emits namespaced console output when enabled
SEARCHABLE: adapter debug, logging hook, adapter utilities
agent-frontmatter:end */

import type { AdapterDebugLogger } from "../create-database-adapter";

export type AdapterDebugOption =
  | boolean
  | {
      logger?: AdapterDebugLogger;
    };

/**
 * Normalises debug logging for adapters. Each adapter provides its options
 * object and receives a composite logger that respects global/on-demand
 * debug settings.
 */
export function createDebugLoggerHook<
  TOptions extends {
    debug?: AdapterDebugOption;
  },
>(prefix: string) {
  return ({
    options,
    defaultLogger,
  }: {
    options: TOptions;
    defaultLogger: AdapterDebugLogger;
  }): AdapterDebugLogger => {
    const debugOption = options.debug;
    if (debugOption === true) {
      return (event, payload) => {
        console.debug(`[${prefix}-adapter] ${event}`, payload);
        defaultLogger(event, payload);
      };
    }
    if (typeof debugOption === "object" && debugOption?.logger) {
      const customLogger = debugOption.logger;
      return (event, payload) => {
        customLogger?.(event, payload);
        defaultLogger(event, payload);
      };
    }
    return defaultLogger;
  };
}
