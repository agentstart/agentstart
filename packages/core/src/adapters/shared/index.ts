/* agent-frontmatter:start
AGENT: Shared adapter helpers
PURPOSE: Centralize reusable adapter utilities
USAGE: import { mapSelectToObject } from "@agent-stack/core/adapters/shared"
EXPORTS: mapSelectToObject
FEATURES:
  - Aggregates shared adapter utilities
  - Keeps helper discovery simple
SEARCHABLE: adapter shared helpers, select helper, adapter utilities
agent-frontmatter:end */

export * from "./debug";
export * from "./map-select";
export * from "./naming";
export * from "./transforms";
export * from "./where";
