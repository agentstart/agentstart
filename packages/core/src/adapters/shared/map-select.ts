/* agent-frontmatter:start
AGENT: Adapter select helpers
PURPOSE: Provide reusable helpers for translating select arrays into driver formats
USAGE: mapSelectToObject(select, model, normalizer, () => true)
EXPORTS: mapSelectToObject
FEATURES:
  - Normalizes field names consistently across adapters
  - Avoids duplicating select-to-object conversion logic
SEARCHABLE: adapter select helper, select normalization, projection helper
agent-frontmatter:end */

/**
 * Converts a string[] selection into a driver-specific object map while applying
 * the adapter's field normalization rules.
 */
export function mapSelectToObject<T>(
  select: string[] | undefined,
  model: string,
  normalizeFieldName: (model: string, field: string) => string,
  mapValue: (field: string) => T,
): Record<string, T> | undefined {
  if (!select?.length) {
    return undefined;
  }
  return Object.fromEntries(
    select.map((field) => [normalizeFieldName(model, field), mapValue(field)]),
  );
}
