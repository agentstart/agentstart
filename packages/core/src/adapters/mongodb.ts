/* agent-frontmatter:start
AGENT: MongoDB adapter
PURPOSE: Provide a database adapter implementation backed by MongoDB
USAGE: mongodbAdapter(db, { provider: "mongodb" })
EXPORTS: mongodbAdapter
FEATURES:
  - Translates normalized filters into MongoDB query documents
  - Supports projections, sorting, and pagination
  - Returns updated documents where supported
SEARCHABLE: mongodb adapter, database adapter, document integration
agent-frontmatter:end */

import { AgentStackError } from "@agent-stack/errors";
import type {
  Db,
  Document,
  Filter,
  FindOneAndDeleteOptions,
  FindOneAndUpdateOptions,
  FindOptions,
  InsertOneResult,
  DeleteResult as MongoDeleteResult,
  UpdateResult as MongoUpdateResult,
  UpdateFilter,
} from "mongodb";

import {
  type AdapterCountArgs,
  type AdapterCreateArgs,
  type AdapterDebugLogger,
  type AdapterDeleteArgs,
  type AdapterFindManyArgs,
  type AdapterFindOneArgs,
  type AdapterUpdateArgs,
  type AdapterUpdateManyArgs,
  createAdapterFactory,
  type DatabaseAdapterMethods,
} from "./create-database-adapter";
import { createDebugLoggerHook } from "./debug";
import { camelToSnake, pluralizeModel } from "./naming";
import {
  type AdapterWhereCondition,
  ensureArrayValue,
  normalizeSortInput,
  normalizeWhereInput,
  splitWhereConditions,
} from "./where";

export interface MongoAdapterUserOptions {
  usePlural?: boolean;
  camelCase?: boolean;
  collectionMap?: Record<string, string>;
  debug?:
    | boolean
    | {
        logger?: AdapterDebugLogger;
      };
}

interface MongoAdapterResolvedOptions extends MongoAdapterUserOptions {
  db: Db;
}

function escapeRegExp(input: string): string {
  return input.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function buildFieldCondition(
  condition: AdapterWhereCondition,
  fieldNormalizer: (field: string) => string,
): Filter<Document> {
  // MongoDB relies on operator-prefixed objects; perform the mapping here so
  // adapters share the same where representation.
  const fieldName = fieldNormalizer(condition.field);
  switch (condition.operator) {
    case "ne":
      return { [fieldName]: { $ne: condition.value } };
    case "lt":
      return { [fieldName]: { $lt: condition.value } };
    case "lte":
      return { [fieldName]: { $lte: condition.value } };
    case "gt":
      return { [fieldName]: { $gt: condition.value } };
    case "gte":
      return { [fieldName]: { $gte: condition.value } };
    case "in": {
      const values = ensureArrayValue(condition);
      return { [fieldName]: { $in: values } };
    }
    case "not_in": {
      const values = ensureArrayValue(condition);
      return { [fieldName]: { $nin: values } };
    }
    case "contains": {
      const pattern = `.*${escapeRegExp(String(condition.value))}.*`;
      return { [fieldName]: { $regex: new RegExp(pattern) } };
    }
    case "starts_with": {
      const pattern = `^${escapeRegExp(String(condition.value))}`;
      return { [fieldName]: { $regex: new RegExp(pattern) } };
    }
    case "ends_with": {
      const pattern = `${escapeRegExp(String(condition.value))}$`;
      return { [fieldName]: { $regex: new RegExp(pattern) } };
    }
    case undefined:
    case "eq":
      return { [fieldName]: condition.value };
    default: {
      const exhaustive: never = condition.operator;
      throw new AgentStackError(
        "DB_ADAPTER_INVALID_WHERE",
        `Unsupported operator "${String(exhaustive)}" supplied for MongoDB adapter.`,
      );
    }
  }
}

function buildMongoFilter(
  conditions: AdapterWhereCondition[],
  fieldNormalizer: (field: string) => string,
): Filter<Document> {
  if (conditions.length === 0) {
    return {};
  }

  const { andConditions, orConditions } = splitWhereConditions(conditions);

  if (andConditions.length === 1 && orConditions.length === 0) {
    return buildFieldCondition(andConditions[0]!, fieldNormalizer);
  }

  const filter: Filter<Document> = {};
  if (andConditions.length > 0) {
    filter.$and = andConditions.map((condition) =>
      buildFieldCondition(condition, fieldNormalizer),
    );
  }
  if (orConditions.length > 0) {
    filter.$or = orConditions.map((condition) =>
      buildFieldCondition(condition, fieldNormalizer),
    );
  }
  return filter;
}

function convertProjection(
  model: string,
  select: string[] | undefined,
  normalizeFieldName: (model: string, field: string) => string,
): FindOptions<Document>["projection"] {
  // Mongo projections expect a document of `{ field: 1 }` pairs.
  if (!select?.length) {
    return undefined;
  }
  return Object.fromEntries(
    select.map((field) => [normalizeFieldName(model, field), 1]),
  );
}

function buildSortDocument(
  model: string,
  sort: AdapterFindManyArgs["sortBy"],
  normalizeFieldName: (model: string, field: string) => string,
): Document | undefined {
  const [primarySort] = normalizeSortInput(sort);
  if (!primarySort?.field) {
    return undefined;
  }
  return {
    [normalizeFieldName(model, primarySort.field)]:
      primarySort.direction === "desc" ? -1 : 1,
  };
}

const baseMongoAdapter = createAdapterFactory<
  MongoAdapterResolvedOptions,
  DatabaseAdapterMethods
>({
  hooks: {
    logger: createDebugLoggerHook<MongoAdapterResolvedOptions>("mongodb"),
    normalizeModelName: ({ options, model }) => {
      const mapped = options.collectionMap?.[model];
      if (mapped) {
        return mapped;
      }
      return options.usePlural ? pluralizeModel(model) : model;
    },
    normalizeFieldName: ({ options, field }) =>
      options.camelCase ? camelToSnake(field) : field,
  },
  config: (options) => ({
    adapterId: "mongodb",
    adapterName: "MongoDB",
    usePlural: options.usePlural ?? false,
    supportsJSON: true,
    supportsDates: true,
    supportsBooleans: true,
    supportsNumericIds: false,
  }),
  adapter: ({
    options,
    wrapOperation,
    normalizeModelName,
    normalizeFieldName,
  }) => {
    const { db } = options;

    const buildFilter = (model: string, where: unknown): Filter<Document> =>
      buildMongoFilter(normalizeWhereInput(where), (field) =>
        normalizeFieldName(model, field),
      );

    return {
      create: wrapOperation<AdapterCreateArgs, unknown>(
        "create",
        async ({ model, data }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const result: InsertOneResult<Document> = await collection.insertOne(
            data as Document,
          );
          const insertedId = result.insertedId;
          if (!insertedId) {
            return data;
          }
          const inserted = await collection.findOne({ _id: insertedId });
          return inserted ?? { ...data, _id: insertedId };
        },
      ),
      findOne: wrapOperation<AdapterFindOneArgs, unknown | null>(
        "findOne",
        async ({ model, where, select }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const projection = convertProjection(
            collectionName,
            select,
            normalizeFieldName,
          );
          return collection.findOne(filter, { projection });
        },
      ),
      findMany: wrapOperation<AdapterFindManyArgs, unknown[]>(
        "findMany",
        async ({ model, where, sortBy, limit, offset }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const cursor = collection.find(filter);
          const sortDocument = buildSortDocument(
            collectionName,
            sortBy,
            normalizeFieldName,
          );
          if (sortDocument) {
            cursor.sort(sortDocument);
          }
          if (typeof offset === "number") {
            cursor.skip(offset);
          }
          if (typeof limit === "number") {
            cursor.limit(limit);
          }
          return cursor.toArray();
        },
      ),
      count: wrapOperation<AdapterCountArgs, number>(
        "count",
        async ({ model, where }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          return collection.countDocuments(filter);
        },
      ),
      update: wrapOperation<AdapterUpdateArgs, unknown>(
        "update",
        async ({ model, where, update }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const updateDoc: UpdateFilter<Document> = {
            $set: update as Document,
          };
          const options: FindOneAndUpdateOptions = {
            returnDocument: "after",
          };
          const result = await collection.findOneAndUpdate(
            filter,
            updateDoc,
            options,
          );
          if (!result) {
            return null;
          }
          return result.value ?? null;
        },
      ),
      updateMany: wrapOperation<AdapterUpdateManyArgs, number>(
        "updateMany",
        async ({ model, where, update }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const updateDoc: UpdateFilter<Document> = {
            $set: update as Document,
          };
          const result: MongoUpdateResult = await collection.updateMany(
            filter,
            updateDoc,
          );
          return result.modifiedCount ?? 0;
        },
      ),
      delete: wrapOperation<AdapterDeleteArgs, unknown>(
        "delete",
        async ({ model, where }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const options: FindOneAndDeleteOptions = {};
          const result = await collection.findOneAndDelete(filter, options);
          if (!result) {
            return null;
          }
          return result.value ?? null;
        },
      ),
      deleteMany: wrapOperation<AdapterDeleteArgs, number>(
        "deleteMany",
        async ({ model, where }) => {
          const collectionName = normalizeModelName(model);
          const collection = db.collection(collectionName);
          const filter = buildFilter(collectionName, where);
          const result: MongoDeleteResult = await collection.deleteMany(filter);
          return result.deletedCount ?? 0;
        },
      ),
      options,
    };
  },
});

export function mongodbAdapter(db: Db, options: MongoAdapterUserOptions = {}) {
  return baseMongoAdapter({
    ...options,
    db,
  });
}
