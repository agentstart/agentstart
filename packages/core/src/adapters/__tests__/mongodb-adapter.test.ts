import type { Db } from "mongodb";
import { describe, expect, it, vi } from "vitest";

import { mongodbAdapter } from "../mongodb";
import { baseAdapterContext } from "./shared/context";
import { runAdapterSuite } from "./shared/run-adapter-suite";

type Operation = { op: string; [key: string]: unknown };

const createStubCollection = (operations: Operation[]) => {
  const chain = {
    sort(sortDoc: unknown) {
      operations.push({ op: "sort", sortDoc });
      return chain;
    },
    skip(skipValue: unknown) {
      operations.push({ op: "skip", skipValue });
      return chain;
    },
    limit(limitValue: unknown) {
      operations.push({ op: "limit", limitValue });
      return chain;
    },
    toArray: async () => {
      operations.push({ op: "toArray" });
      return [{ id: "result" }];
    },
  };

  return {
    insertOne: async (doc: unknown) => {
      operations.push({ op: "insertOne", doc });
      return { insertedId: "inserted" };
    },
    findOne: async (filter: unknown, options?: unknown) => {
      operations.push({ op: "findOne", filter, options });
      return { id: "found" };
    },
    find: (filter: unknown, options?: unknown) => {
      operations.push({ op: "find", filter, options });
      return chain;
    },
    countDocuments: async (filter: unknown) => {
      operations.push({ op: "countDocuments", filter });
      return 7;
    },
    findOneAndUpdate: async (
      filter: unknown,
      update: unknown,
      options?: unknown,
    ) => {
      operations.push({ op: "findOneAndUpdate", filter, update, options });
      return { value: { updated: true } };
    },
    updateMany: async (filter: unknown, update: unknown) => {
      operations.push({ op: "updateMany", filter, update });
      return { modifiedCount: 3 };
    },
    findOneAndDelete: async (filter: unknown) => {
      operations.push({ op: "findOneAndDelete", filter });
      return { value: { deleted: true } };
    },
    deleteMany: async (filter: unknown) => {
      operations.push({ op: "deleteMany", filter });
      return { deletedCount: 2 };
    },
  };
};

function buildMongoInstance(overrides?: {
  getFieldAttributes?: (
    model: string,
    field: string,
  ) => Record<string, unknown>;
}) {
  const ops: Operation[] = [];
  const collection = createStubCollection(ops);
  const db = {
    collection: vi.fn(() => collection),
  } as unknown as Db;
  const adapter = mongodbAdapter(db, { camelCase: true, usePlural: true });
  const context = {
    ...baseAdapterContext,
    ...(overrides?.getFieldAttributes
      ? { getFieldAttributes: overrides.getFieldAttributes }
      : {}),
  };
  const methods = adapter.initialize(context);
  return { methods, ops };
}

runAdapterSuite({
  name: "mongodbAdapter",
  createAdapter: () => buildMongoInstance().methods,
  scenario: {
    create: { model: "user", data: { firstName: "Ada" } },
    findOne: {
      model: "user",
      where: [{ field: "firstName", value: "Ada" }],
      select: ["firstName"],
    },
    findMany: {
      model: "user",
      where: [
        { field: "status", value: "active" },
        { field: "tags", value: ["beta"], operator: "in", connector: "OR" },
      ],
      sortBy: [{ field: "createdAt", direction: "desc" }],
      limit: 5,
      offset: 1,
    },
    count: { model: "user" },
    update: {
      model: "user",
      where: [{ field: "id", value: "1" }],
      update: { status: "inactive" },
    },
    updateMany: {
      model: "user",
      where: [{ field: "status", value: "inactive" }],
      update: { status: "active" },
    },
    delete: {
      model: "user",
      where: [{ field: "id", value: "1" }],
    },
    deleteMany: {
      model: "user",
      where: [{ field: "status", value: "active" }],
    },
    upsert: {
      model: "user",
      where: [{ field: "id", value: "1" }],
      create: { id: "1", status: "active" },
      update: { status: "pending" },
      select: ["status"],
    },
  },
  assertions: {
    create: (record) => expect(record).toMatchObject({ id: "found" }),
    findOne: (record) => expect(record).toEqual({ id: "found" }),
    findMany: (records) => expect(records).toEqual([{ id: "result" }]),
    count: (value) => expect(value).toBe(7),
    update: (record) => expect(record).toEqual({ updated: true }),
    updateMany: (value) => expect(value).toBe(3),
    delete: (record) => expect(record).toEqual({ deleted: true }),
    deleteMany: (value) => expect(value).toBe(2),
    upsert: (record) => expect(record).toEqual({ updated: true }),
  },
});

describe("mongodbAdapter", () => {
  it("translates filters, projections, and write operations", async () => {
    const { ops, methods } = buildMongoInstance();

    await methods.create({ model: "user", data: { firstName: "Ada" } });
    await methods.findOne({
      model: "user",
      where: [{ field: "firstName", value: "Ada" }],
      select: ["firstName"],
    });
    await methods.findMany({
      model: "user",
      where: [
        { field: "status", value: "active" },
        { field: "tags", value: ["beta"], operator: "in", connector: "OR" },
      ],
      sortBy: [{ field: "createdAt", direction: "desc" }],
      limit: 5,
      offset: 1,
    });
    await methods.count({ model: "user" });
    await methods.update({
      model: "user",
      where: [{ field: "id", value: "1" }],
      update: { status: "inactive" },
    });
    await methods.updateMany({
      model: "user",
      where: [{ field: "status", value: "inactive" }],
      update: { status: "active" },
    });
    await methods.delete({
      model: "user",
      where: [{ field: "id", value: "1" }],
    });
    await methods.deleteMany({
      model: "user",
      where: [{ field: "status", value: "active" }],
    });
    const upserted = await methods.upsert({
      model: "user",
      where: [{ field: "id", value: "1" }],
      create: { id: "1", status: "active" },
      update: { status: "pending" },
      select: ["status"],
    });
    expect(upserted).toEqual({ updated: true });

    expect(ops).toMatchSnapshot();
  });

  it("applies default field attributes on insert", async () => {
    const defaults = (model: string, field: string) => {
      if (model === "user" && field === "__fields__") {
        return { status: true };
      }
      if (model === "user" && field === "status") {
        return { defaultValue: "pending" };
      }
      return {};
    };
    const { ops, methods } = buildMongoInstance({
      getFieldAttributes: defaults,
    });

    await methods.create({
      model: "user",
      data: { firstName: "Grace" },
    });

    const insertOperation = ops.find((entry) => entry.op === "insertOne");
    expect(insertOperation?.doc).toMatchObject({ status: "pending" });
  });
});
