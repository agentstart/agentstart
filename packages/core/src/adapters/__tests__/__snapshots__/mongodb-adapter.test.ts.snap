// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mongodbAdapter > translates filters, projections, and write operations 1`] = `
[
  {
    "doc": {
      "firstName": "Ada",
    },
    "op": "insertOne",
  },
  {
    "filter": {
      "_id": "inserted",
    },
    "op": "findOne",
    "options": undefined,
  },
  {
    "filter": {
      "first_name": "Ada",
    },
    "op": "findOne",
    "options": {
      "projection": {
        "first_name": 1,
      },
    },
  },
  {
    "filter": {
      "$and": [
        {
          "status": "active",
        },
      ],
      "$or": [
        {
          "tags": {
            "$in": [
              "beta",
            ],
          },
        },
      ],
    },
    "op": "find",
    "options": undefined,
  },
  {
    "op": "sort",
    "sortDoc": {
      "created_at": -1,
    },
  },
  {
    "op": "skip",
    "skipValue": 1,
  },
  {
    "limitValue": 5,
    "op": "limit",
  },
  {
    "op": "toArray",
  },
  {
    "filter": {},
    "op": "countDocuments",
  },
  {
    "filter": {
      "id": "1",
    },
    "op": "findOneAndUpdate",
    "options": {
      "returnDocument": "after",
    },
    "update": {
      "$set": {
        "status": "inactive",
      },
    },
  },
  {
    "filter": {
      "status": "inactive",
    },
    "op": "updateMany",
    "update": {
      "$set": {
        "status": "active",
      },
    },
  },
  {
    "filter": {
      "id": "1",
    },
    "op": "findOneAndDelete",
  },
  {
    "filter": {
      "status": "active",
    },
    "op": "deleteMany",
  },
  {
    "filter": {
      "id": "1",
    },
    "op": "findOneAndUpdate",
    "options": {
      "projection": {
        "status": 1,
      },
      "returnDocument": "after",
      "upsert": true,
    },
    "update": {
      "$set": {
        "status": "pending",
      },
      "$setOnInsert": {
        "id": "1",
        "status": "active",
      },
    },
  },
]
`;
